### PR TITLE
datetime: Set the default year to the year during compile

### PR DESCRIPTION
**Pull request by FintasticMan**
Supersedes #2067.

### CODE REVIEW

**Comment by tituscmd** [comment]
While were at it changing stuff related to Date & Time, we could also consider this PR:

https://github.com/InfiniTimeOrg/InfiniTime/issues/1782

**Review by NeroBurner — APPROVED** [review]
Code looks good. When tested on device good to go

File: src/components/datetime/DateTimeController.cpp [hunk 1]
```diff
@@ -7,16 +7,28 @@
...
+    return result;
+  }
 }
 
 DateTime::DateTime(Controllers::Settings& settingsController) : settingsController {settingsController} {
   mutex = xSemaphoreCreateMutex();
   ASSERT(mutex != nullptr);
   xSemaphoreGive(mutex);
+
+  SetTime(compileTimeAtoi(&__DATE__[7]), 0, 0, 0, 0, 0);
```
Comment (mark9064) [inline src/components/datetime/DateTimeController.cpp:31]: Maybe add a comment explaining char 7->end encodes the year? Not a huge deal though

**Review by mark9064 — APPROVED** [review]
LGTM. Haven't tested on HW

PS: Took me about 5 minutes to understand the const correctness of the arrays, but I think you've got the type right. Well done 😅

**Comment by FintasticMan** [comment]
Yeah, const-correctness in C++ isn't very obvious haha. I'm not by my dev kit so I can't debug this, but it really doesn't seem to like calling `SetTime` in the constructor, as doing so seems to result in a non-booting firmware.

**Comment by mark9064** [comment]
Ah, that would be as system task doesn't exist yet so it hits a null ptr deref when it tries to push OnNewTime

**Comment by FintasticMan** [comment]
Good thinking! Adding a NULL check should indeed fix it.

**Comment by FintasticMan** [comment]
It boots now, but I haven't been able to get the time to not be read from `.noinit` :sweat_smile:, so I can't test if it actually works. I'm sure if the battery runs out it'll reset the time (isn't it annoying having such good battery life!)

**Comment by NeroBurner** [comment]
Is it possible to make a Firmware that always reads the noinit data I instead of the filesystem?

**Comment by mark9064** [comment]
image
So I tried @pipe01's InfiniEmu and it looks like we're off by a little bit. I'd guess the month and day are meant to be ones

**Comment by FintasticMan** [comment]
That makes a lot of sense haha.

**Comment by FintasticMan** [comment]
My watch just ran out of battery, so I can confirm: this does work on actual hardware!

File: src/components/datetime/DateTimeController.cpp [hunk 2]
```diff
@@ -46,7 +59,9 @@ void DateTime::SetTime(uint16_t year, uint8_t month, uint8_t day, uint8_t hour,
   UpdateTime(previousSystickCounter, true);
   xSemaphoreGive(mutex);
 
-  systemTask->PushMessage(System::Messages::OnNewTime);
+  if (systemTask) {
```
Comment (mark9064) [inline src/components/datetime/DateTimeController.cpp:62]: The other checks in the file do `systemTask != nullptr`, maybe worth being consistent? Very minor though

**Review by mark9064 — APPROVED** [review]

**Review by JF002 — APPROVED** [review]

**Pull request merged**